app/models/katello/concerns/host_managed_extensions.rb

```diff
@@ -326,26 +326,77 @@ def import_enabled_repositories(repos)
-        end
-        sync_available_module_stream_associations(streams)
-      end
-
-      def sync_available_module_stream_associations(new_available_module_streams)
+        Rails.logger.debug "INSERT_ALL-------------------------------------"
+        Rails.logger.debug module_streams.first(3)
+        streams = module_streams.map do |module_stream|
+          {
+            name: module_stream["name"],
+            stream: module_stream["stream"],
+            context: module_stream["context"]
+          }
+        end
+        result = AvailableModuleStream.upsert_all(
+          streams,
+          unique_by: %w[name stream context],
+          returning: %w[id name stream context]
+        )
+        Rails.logger.debug "RESULT-------------------------------------"
+        Rails.logger.debug result.to_a
+        # original module_streams looks like this
+        # {"name"=>"389-ds", "stream"=>"1.4", "version"=>"8030020201203210520", "context"=>"e114a9e7", "arch"=>"x86_64", "profiles"=>[], "installed_profiles"=>[], "status"=>"default", "active"=>false}
+        
+        # now we transform it into this with Hash#hash
+        # { 4156191055189065161 => {"name"=>"389-ds", "stream"=>"1.4", "version"=>"8030020201203210520", "context"=>"e114a9e7", "arch"=>"x86_64", "profiles"=>[], "installed_profiles"=>[], "status"=>"default", "active"=>false} }
+        indexed_module_streams = module_streams.index_by { |module_stream| module_stream.slice("name", "stream", "context").hash }
+        
+        # result.to_a looks like this
+        # [{"id"=>1, "name"=>"389-ds", "stream"=>"1.4", "context"=>"e114a9e7"}, {"id"=>2, "name"=>"389-ds", "stream"=>"1.4", "context"=>"e114a9e7"}]
+        # now we transform it into this with Hash#hash
+        # { 4156191055189065161 => {"id"=>1, "name"=>"389-ds", "stream"=>"1.4", "context"=>"e114a9e7"}, 4156191055189065162 => {"id"=>2, "name"=>"389-ds", "stream"=>"1.4", "context"=>"e114a9e7"} }
+        indexed_result = result.to_a.index_by { |row| row.slice("name", "stream", "context").hash }
+        Rails.logger.debug "INDEXED_RESULT-------------------------------------"
+        Rails.logger.debug indexed_result
+
+        untransformed_keys = indexed_module_streams.keys
+        Rails.logger.debug "UNTRANSFORMED_KEYS-------------------------------------"
+        Rails.logger.debug untransformed_keys
+        # now we transform the keys of indexed_module_streams to the ids from indexed_result
+        # { 1 => {"name"=>"389-ds", "stream"=>"1.4", "version"=>"8030020201203210520", "context"=>"e114a9e7", "arch"=>"x86_64", "profiles"=>[], "installed_profiles"=>[], "status"=>"default", "active"=>false} }
+        indexed_module_streams.transform_keys! { |k| indexed_result[k]&.[]("id") }
+
+        Rails.logger.debug "TRANSFORMED KEYS-------------------------------------"
+        Rails.logger.debug indexed_module_streams.keys
+        # remove the keys that were not transformed
```

Comment:
Maybe I'm missing something, but what do these hash transformations get us that `result` doesn't already have? The only module streams that will have DB IDs will be the ones returned by `upsert_all`, which are all of the records in `result`.

As I understand the the code here, `result` and `module_streams` get matched up by the hash of their (name, stream, context) set.  Then, the hashes get replaced by DB IDs if they exist. However, `result` already has these database IDs, and `indexed_module_streams.except!(untransformed_keys)` throws out all of the ones that don't have database IDs.

 Doesn't this just leave us with what is in `result` but in a slightly different format?

I see that `indexed_module_streams` will also get you the profiles, but that could be returned from `upsert_all`.

Reply:
> Maybe I'm missing something, but what do these hash transformations get us that `result` doesn't already have?

`result` is an ActiveRecord::Result and thus can only give us columns that are in our tables; in other words, attributes of the AvailableModuleStream model. However, the argument passed into this method contains several fields that are _not_ attributes of that model. The other difference is that `result` contains _only_ newly-created records, not existing records.   So we have

`result` - newly inserted Katello DB records, containing only id, name, stream, context
`module_streams` - module stream data containing name, stream, version, context, arch, profiles, installed_profiles, status, active -- but NOT id

The next method, `sync_available_module_stream_associations`, in its original form, expected an indexed Hash by database ID containing _both_ existing _and_ newly created AvailableModuleStreams. It called this `new_available_module_streams` even though it contained both existing and new records. It used this in two ways:

1. Composing a list of AvailableModuleStreams to delete, by comparing the host's `available_module_stream_ids` with `new_available_module_streams.keys`
2. Composing lists of new and upgradable module streams, by using `new_available_module_streams.keys`

The difficult part is that `sync_available_module_stream_associations` relies on our hash meeting _both_ of the following requirements:

1. Indexed by Katello database ID; and
2. contains _both_ new records _and_ existing records.

Now that I think some more, it may be simpler to just look at the database after doing our `upsert_all`, and not rely anymore on the passed-in module stream data.

Reply:
update: Refactored to index the hash using DB queries, see what you think